streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java

```diff
@@ -36,7 +36,7 @@
     private final TaskId taskId;
     private final String applicationId;
     private final StreamsConfig config;
-    private final StreamsMetrics metrics;
+    private final StreamsMetricsImpl metrics;
```

Comment:
Using the interface is really only useful in our public ProcessorContext interface. Using the StreamsMetrics interface in our internals just forces us to cast it back to StreamsMetricsImpl all over the place.

I've changed it to S.M.I. here and elsewhere to cut down on the casting. The only things that should need to cast now are components that get the metrics via ProcessorContext, and they should always perform that cast as early as possible to prevent post-initialization runtime exceptions.